scripts/DllManipulatorScript.cs

```diff
@@ -3,10 +3,12 @@
 using System.Threading;
 using System.Linq;
 using UnityEngine;
+using UnityEditor;
```

Comment:
I assume that you mean to merge the two files `DllManipulatorScript` and `DllManipulatorEditor`. Would it not make more sense to just move the `DllManipulatorScript` to the `Editor` folder and keep them separate?

Reply:
I don't really want to move the main executable script to Editor folder, because I want it to be able to run it outside editor. I know I said it isn't recommended to  publish games with that way, but a quick Build and Run won't hurt.

What I'm thinking now is to move most of the code from  `DllManipulatorScript` to `DllManipulator` (which should really be done anyway) and make separate, dedicated script in Editor folder (or maybe use existing `DllManipulatorEditor`?). Btw, does a script with `[ExecuteInEditMode]` have to be somehow instantiated in order to be run? I will look at this more tomorrow.

Reply:
Makes sense. I'd personally keep the custom editor IMGUI code separate.

Yes there has to be an instance for `[ExecuteInEditMode]`.
> Makes _all instances_ of a script execute in Edit Mode (Unity docs)

Reply:
Do you maybe know if there is a way to bootstrap code in editor, without a script in the scene? Because now that would require instantiating two scripts, one 'normal' and one for editor, with [ExecuteInEditMode]. I remember searching for something like this some time ago, but I think I didn't find anything suitable. For instance, a constructor of a custom editor class get's called automatically, but only when you select the target script. And if there was a way to do that, I could potentially even cease requiring putting my script in the scene at all.

Reply:
From what I know you always need an instance in the scene (for it to work without any additional scripting by the user). Theres only [`        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.BeforeSplashScreen)]
`](https://docs.unity3d.com/ScriptReference/RuntimeInitializeOnLoadMethodAttribute.html?_ga=2.201383251.333741474.1584180091-526827772.1567673391) but that would only be useful when using this tool in play mode.

I don't see why you would _need_ two separate scripts, you can use `Application.isPlaying` and/or `#if UNITY_EDITOR` in the same script. 

I guess you can use `RuntimeInitializeOnLoadMethod` for the 'normal' script and add an instance to the scene when you want to use it in edit mode as well? However, why would you not always use the edit mode version (given all the editor parts are in `#if UNITY_EDITOR` so you can still use it in a build)?

Reply:
Because scripts inside Editor folder are compiled into `Assembly-CSharp-Editor` assembly and these outside of Editor folder are compiled into  `Assembly-CSharp` and only the former references `UnityEditor.dll` assembly, so you can't access the `UnityEditor` namespace in the non-editor scripts, even when `UNITY_EDITOR` symbol is specified. You can have custom assembly definition files that alter this behavior, but I don't want to force users to use them.

So that's how I remember it working, but now that I check it again I see that `Assembly-CSharp` also references `UnityEditor.dll`. So unless I messed this up, this behavior has changed in some version of unity. And if that's the case, I want be stay compatible with previous versions and not rely on this change.

As to the `[RuntimeInitializeOnLoadMethod]`, attribute:
>Methods marked [RuntimeInitializeOnLoadMethod] are invoked after the game has been loaded. This is after the Awake method has been invoked.
Note: The execution order of methods marked `[RuntimeInitializeOnLoadMethod]` is not guaranteed.

Which means that I can't safety use it, because some user code that uses a DLL could be have been run earlier, which would cause the DLL to be loaded by Unity (mono to be specific), and disallow further unloading.

Reply:
Fair point, however, I understood it that the Editor folders are just to save you from putting `#if`s and you can still freely access the `UnityEditor` namespace outside of such a folder. Files in an `Editor` folder are not included in a build. So I think it should be fine if we use `#if`s and put the script outside the `Editor` folder. (citing [docs](https://docs.unity3d.com/Manual/SpecialFolders.html) and [this](http://answers.unity.com/answers/993653/view.html) from 2015)

Regarding the attribute, I think that's because the _default load type_ is after the scene has been loaded. See decompiled ctor below.
```cs
public RuntimeInitializeOnLoadMethodAttribute()
{
  this.loadType = RuntimeInitializeLoadType.AfterSceneLoad;
}
```
So if you specify the `BeforeSplashScreen` or `BeforeSceneLoad` [LoadType](https://docs.unity3d.com/ScriptReference/RuntimeInitializeLoadType.html) it might work as intended.

Reply:
I've checked the Unity 2018 and your're right, although not all UnityEditor assemblies are referenced, the core one is and can be accessed. So this may freely stay how you initially proposed, but with `#if UNITY_EDITOR` checks.